library/core/registry/src/main/java/org/quiltmc/qsl/registry/mixin/BlocksMixin.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2021-2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.registry.mixin;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.Blocks;
+import net.minecraft.util.registry.Registry;
+import org.quiltmc.qsl.registry.api.event.RegistryMonitor;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(Blocks.class)
+public abstract class BlocksMixin {
+	@Inject(method = "<clinit>", at = @At("RETURN"))
+	private static void onInit(CallbackInfo ci) {
+		RegistryMonitor.create(Registry.BLOCK)
+				.forUpcoming(context -> context.value().getStateManager().getStates().forEach(Block.STATE_IDS::add));
```

Comment:
If you use forUpcoming you can use the event directly?

Not sure which way should be the default.